### PR TITLE
Support codegen for std, std.dim, std.correction

### PR DESCRIPTION
**Pull request by wconstab**
#65576

### CODE REVIEW

File: tools/codegen/dest/lazy_ir.py [hunk 1]
```diff
@@ -29,6 +29,9 @@ def node_ctor_inputs(func: LazyIrSchema) -> str:
         else:
             if isinstance(arg.type, VectorCType) and isinstance(arg.type.elem, BaseCType):
                 node_ctor_values.append(f"lazy_tensors::util::ToVector<{arg.type.elem.type}>({arg.name})")
+            elif (isinstance(arg.type, OptionalCType) and
+                  isinstance(arg.type.elem, VectorCType)):
+                node_ctor_values.append(f"lazy_tensors::util::ToOptionalVector<{arg.type.elem.elem.type}>({arg.name})")
```
Comment (wconstab) [inline tools/codegen/dest/lazy_ir.py:34]: @bdhirsh am I doing this in the least awkward way possible?

Comment (bdhirsh) [reply tools/codegen/dest/lazy_ir.py:34]: I think... this is okay haha.

One easy to clean it up a little bit would be to have a function that operates on individual arguments, e.g. from `CType -> str`. Then your funcgtion that maps from `LazyIrSchema` -> "stuff that goes into the constructor" doesn't have to care about what the per-argument mapping looks like.

Another thing that would definitely be more involved, so I'm not sure that it warrants doing now - Ed also gave me a lot of high-level codegen feedback on my recent PR's that probably applies here. `GenLazyNativeFuncDefinition.__call__()` does a lot of stuff and has a lot of string parsing lying around. One example of how to factor the logic some more in this specific case - calling the `Node` constructor - is to write the call as a translation between API's:

```
# this is the kernel that gets registered to the dispatcher
# It has an API corresponding to the dispatcher API
sig = kernel_signature(func, self.backend_index)

# This object knows how to declare/define LazyTensor node constructors
# It follows a different API compared to the dispatcher. For example, it only involves value types.
# And it also requires that some inputs (like vectors and optionals) be transformed by some utils first.
node_ctr_sig = LazyNodeSignature(func) 
{sig.decl(name=f"{self.class_method_name}::{schema.aten_name}")} {{

auto node = node_ctr_sig.inner_call()
}};
```
I ended up with something similar in my functionalization pass codegen [here](https://github.com/pytorch/pytorch/blob/a510f4139b2de1f5baed13bd50a035e39fd2a78b/tools/codegen/api/types.py#L546). Then again, maybe Ed will have some more feedback on it when he gets back :)

File: lazy_tensor_core/lazy_tensor_core/csrc/LazyShapeDtype.cpp [hunk 2]
```diff
@@ -43,14 +43,51 @@
...
+  return compute_shape_std(self, c10::nullopt, c10::nullopt, false);
+}
+std::vector<std::vector<int64_t>> compute_shape_std(const at::Tensor & self, at::IntArrayRef dim, bool unbiased, bool keepdim){
+  return compute_shape_std(self, dim, c10::nullopt, keepdim);
+}
+std::vector<std::vector<int64_t>> compute_shape_std(const at::Tensor & self, c10::optional<at::IntArrayRef> dim, c10::optional<int64_t> correction, bool keepdim){
+  if (dim.has_value()) {
+    auto sizes = self.sizes().vec();
+    if (keepdim) {
+      for (auto reduce_dim:  dim.value()) {
```
Comment (alanwaketan) [inline lazy_tensor_core/lazy_tensor_core/csrc/LazyShapeDtype.cpp:64]: Remove one space after :.

File: lazy_tensor_core/lazy_tensor_core/csrc/LazyShapeDtype.cpp [hunk 3]
```diff
@@ -43,14 +43,51 @@
...
+        sizes[at::maybe_wrap_dim(reduce_dim, self)] = 1;
+      }
+      return {sizes};
+    } else {
+      std::vector<int64_t> reduce_dims(dim.value().begin(), dim.value().end());
+      at::maybe_wrap_dims(reduce_dims, self.sizes().size());
+      sort(reduce_dims.begin(), reduce_dims.end(), std::greater<int64_t>());
+      for (auto reduce_dim:  reduce_dims) {
```
Comment (alanwaketan) [inline lazy_tensor_core/lazy_tensor_core/csrc/LazyShapeDtype.cpp:72]: Ditto.

File: lazy_tensor_core/lazy_tensor_core/csrc/LazyShapeDtype.cpp [hunk 4]
```diff
@@ -43,14 +43,51 @@
...
+  if (dim.has_value()) {
+    auto sizes = self.sizes().vec();
+    if (keepdim) {
+      for (auto reduce_dim:  dim.value()) {
+        sizes[at::maybe_wrap_dim(reduce_dim, self)] = 1;
+      }
+      return {sizes};
+    } else {
+      std::vector<int64_t> reduce_dims(dim.value().begin(), dim.value().end());
+      at::maybe_wrap_dims(reduce_dims, self.sizes().size());
```
Comment (alanwaketan) [inline lazy_tensor_core/lazy_tensor_core/csrc/LazyShapeDtype.cpp:70]: Just `self` is fine?

File: lazy_tensor_core/lazy_tensor_core/csrc/LazyShapeDtype.cpp [hunk 5]
```diff
@@ -43,14 +43,51 @@
...
+      sort(reduce_dims.begin(), reduce_dims.end(), std::greater<int64_t>());
+      for (auto reduce_dim:  reduce_dims) {
+        sizes.erase(sizes.begin() + reduce_dim);
```
Comment (alanwaketan) [inline lazy_tensor_core/lazy_tensor_core/csrc/LazyShapeDtype.cpp:73]: Why don't you return the `sizes`?

File: lazy_tensor_core/lazy_tensor_core/csrc/LazyShapeDtype.cpp [hunk 6]
```diff
@@ -43,14 +43,51 @@
...
+std::vector<std::vector<int64_t>> compute_shape_std(const at::Tensor & self, c10::optional<at::IntArrayRef> dim, c10::optional<int64_t> correction, bool keepdim){
+  if (dim.has_value()) {
+    auto sizes = self.sizes().vec();
+    if (keepdim) {
+      for (auto reduce_dim:  dim.value()) {
+        sizes[at::maybe_wrap_dim(reduce_dim, self)] = 1;
+      }
+      return {sizes};
+    } else {
+      std::vector<int64_t> reduce_dims(dim.value().begin(), dim.value().end());
```
Comment (alanwaketan) [inline lazy_tensor_core/lazy_tensor_core/csrc/LazyShapeDtype.cpp:69]: @bdhirsh Mind take a look on this part? I'm not sure.

File: lazy_tensor_core/lazy_tensor_core/csrc/LazyShapeDtype.cpp [hunk 7]
```diff
@@ -43,14 +43,51 @@
  *
  */
 
+#include <functional>
```
Comment (alanwaketan) [inline lazy_tensor_core/lazy_tensor_core/csrc/LazyShapeDtype.cpp:46]: Why is this needed?

File: lazy_tensor_core/lazy_tensor_core/csrc/LazyShapeDtype.cpp (hunk 5 above)
Comment (wconstab) [reply lazy_tensor_core/lazy_tensor_core/csrc/LazyShapeDtype.cpp:72]: oh no! that was ugly. fixed.

File: lazy_tensor_core/lazy_tensor_core/csrc/LazyShapeDtype.cpp (hunk 4 above)
Comment (wconstab) [reply lazy_tensor_core/lazy_tensor_core/csrc/LazyShapeDtype.cpp:69]: you mean 
at::maybe_wrap_dims(reduce_dims, self) ?

File: lazy_tensor_core/lazy_tensor_core/csrc/LazyShapeDtype.cpp (hunk 3 above)
Comment (wconstab) [reply lazy_tensor_core/lazy_tensor_core/csrc/LazyShapeDtype.cpp:71]: What is this referring to?

File: lazy_tensor_core/lazy_tensor_core/csrc/LazyShapeDtype.cpp (hunk 4 above)
Comment (alanwaketan) [reply lazy_tensor_core/lazy_tensor_core/csrc/LazyShapeDtype.cpp:69]: Yea.

File: lazy_tensor_core/lazy_tensor_core/csrc/LazyShapeDtype.cpp (hunk 3 above)
Comment (alanwaketan) [reply lazy_tensor_core/lazy_tensor_core/csrc/LazyShapeDtype.cpp:71]: It should be `for (auto reduce_dim : reduce_dims) {`?

Comment (wconstab) [reply lazy_tensor_core/lazy_tensor_core/csrc/LazyShapeDtype.cpp:71]: oh.. the space. i thought you meant something about 'self'

Comment (alanwaketan) [reply lazy_tensor_core/lazy_tensor_core/csrc/LazyShapeDtype.cpp]: Oh, I should pay more attentions to the actual order of comments being printed in the GH UI.

File: lazy_tensor_core/lazy_tensor_core/csrc/LazyShapeDtype.cpp (hunk 4 above)
Comment (wconstab) [reply lazy_tensor_core/lazy_tensor_core/csrc/LazyShapeDtype.cpp:69]: oh. no, it doesn't work that way.  maybe_wrap_dims is unfortunately not defined in the same way as maybe_wrap_dim.  It's defined like this:

```
template <typename Container>
inline void maybe_wrap_dims(Container& dims, int64_t dim_post_expr) {
  return maybe_wrap_dims_n(dims.data(), dims.size(), dim_post_expr);
}
```

Comment (alanwaketan) [reply lazy_tensor_core/lazy_tensor_core/csrc/LazyShapeDtype.cpp:69]: Sad, then maybe `self.dim()`?

Comment (wconstab) [reply lazy_tensor_core/lazy_tensor_core/csrc/LazyShapeDtype.cpp:69]: ahh yes. better. thanks!

File: tools/codegen/dest/lazy_ir.py (hunk 1 above)
Comment (wconstab) [reply tools/codegen/dest/lazy_ir.py:34]: OK. I at least refactored following the first (simpler) suggestion for now.

File: tools/codegen/dest/lazy_ir.py [hunk 8]
```diff
@@ -10,30 +10,36 @@
...
+            return f"lazy_tensors::util::ToVector<{arg.type.elem.type}>({arg.name})"
+        elif (isinstance(arg.type, OptionalCType) and
+                isinstance(arg.type.elem, VectorCType)):
+            return f"lazy_tensors::util::ToOptionalVector<{arg.type.elem.elem.type}>({arg.name})"
+        else:
+            return f"{arg.name}"
 
-    node_ctor_inputs_str = ",\n                              ".join(node_ctor_values)
-    return node_ctor_inputs_str
+def node_ctor_inputs(func: LazyIrSchema) -> str:
```
Comment (bdhirsh) [inline tools/codegen/dest/lazy_ir.py:37]: I think it would maybe be more readable to couple this with the actual call-site. Since you have two tightly coupled pieces of code (the arguments to the node constructor, and the name of the constructor itself) that are very non-local. Maybe this should just return the entire string:
```
return f'torch::lazy::MakeNode<ir::ops::{func.node_name}>({node_ctr_args})'
```

Comment (wconstab) [reply tools/codegen/dest/lazy_ir.py:37]: yea that's a good point. i'll clean it up.

File: lazy_tensor_core/lazy_tensor_core/csrc/LazyShapeDtype.cpp (hunk 6 above)
Comment (bdhirsh) [reply lazy_tensor_core/lazy_tensor_core/csrc/LazyShapeDtype.cpp:68]: oh okay - this one's more complicated because it uses `TensorIterator::make_reduction()` in core. Maybe that would be a good candidate to port to structured instead of trying to replicate the main logic here 😛 .

I'm happy to take a look / try to confirm if you have most of the logic. It's a little unfortunate that we don't have an automated way to check their correctness for now. PyTorch/XLA runs our device-agnostic tests in core on their backend - so once they start using these formulas then they'll get tested. Although I wonder if we can get the same testing for the `Lazy` devices sooner.

File: lazy_tensor_core/lazy_tensor_core/csrc/LazyShapeDtype.cpp [hunk 9]
```diff
@@ -43,14 +43,51 @@
...
+  if (dim.has_value()) {
+    auto sizes = self.sizes().vec();
+    if (keepdim) {
+      for (auto reduce_dim: dim.value()) {
+        sizes[at::maybe_wrap_dim(reduce_dim, self)] = 1;
+      }
+    } else {
+      std::vector<int64_t> reduce_dims(dim.value().begin(), dim.value().end());
+      at::maybe_wrap_dims(reduce_dims, self.dim());
+      sort(reduce_dims.begin(), reduce_dims.end(), std::greater<int64_t>());
```
Comment (bdhirsh) [inline lazy_tensor_core/lazy_tensor_core/csrc/LazyShapeDtype.cpp:70]: Where does this sorting logic come from? I'm staring at the code for `make_reduction()`, and the case where `keepdim=False` seems pretty similar to the `true` case ([here](https://github.com/pytorch/pytorch/blob/0bd0a67c4fce0ff9f005ae38d10df5ac778e7a54/aten/src/ATen/native/ReduceOpsUtils.h#L134)). Something like:
```
  auto shape = DimVector(self.sizes());
  for (int dim = shape.size() - 1; dim >= 0; dim--) {
    if (mask[dim]) {
      if (keepdim) {
        shape[dim] = 1;
      } else {
        shape.erase(shape.begin() + dim);
      }
    }
  }
  return shape;
```

Comment (wconstab) [reply lazy_tensor_core/lazy_tensor_core/csrc/LazyShapeDtype.cpp:70]: ahh. i should have used that code, i'll change to use it. 

In their version they construct a 'dimmask' which is equivalent to my sorting, in terms of canonicalizing the keepdims in some order, so that you can safely remove them from the vector starting from the higher indices toward the lower indicies.

The issue would be if you tried to remove dim 1 then later remove dim 3, you'd have to keep track of the indices above the thing you remove all shifting down.

**Review by bdhirsh — APPROVED** [review]
I'm hoping we can figure out some way to run LazyTensor on the backend-agnostic test suite in core to help uncover any problems in the hand-written shape inference code, but this seems ok to me.
